File: svws-core/src/main/java/de/nrw/schule/svws/core/data/kalender/KalenderEintrag.java
package de.nrw.schule.svws.core.data.kalender;

import jakarta.xml.bind.annotation.XmlRootElement;

import de.nrw.schule.svws.core.transpiler.TranspilerDTO;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.constraints.NotNull;

/**
 * Ein Eintrag in einem Kalender
 */
@XmlRootElement
@Schema(description = "Ein Eintrag in einem Kalender.")
@TranspilerDTO
public class KalenderEintrag {
	/** ID des Kalendereintrags */
	@Schema(required = true, description = "die ID des Kalendereintrag", example = "123")
	public @NotNull String id = "";

	/** ID des Kalenders */
	@Schema(required = true, description = "die ID des Kalenders, zu dem der Kalendereintrag gehört", example = "global")
	public @NotNull String kalenderId = "";

	/**
	 * URI der ICS-Repräsentation des Kalendereintrags
	 */
	@Schema(required = true, description = "uid des kalenderEintrags", example = "exf80")
	public @NotNull String uid = "";

	/**
	 * Versionskennzeichen des Kontaks
	 */
	@Schema(required = true, description = "Versionskennzeichen des Kontakts", example = "78")
	public @NotNull String version = "";

	/** das serialisierte .ics dieses Kalendereintrags */
	@Schema(required = true, description = "das serialisierte .ics dieses Kalendereintrags", example = "...")
	public @NotNull String data = "";

	/** der Startzeitpunkt dieses Kalendereintrags als SQL-Timestamp */
	@Schema(required = false, description = "der Startzeitpunkt dieses Kalendereintrags ", example = "2022-10-13 14:00:00.000")
	public String kalenderStart;

	/** der Endzeitpunkt dieses Kalendereintrags als SQL-Timestamp */
	@Schema(required = false, description = "der Endzeitpunkt dieses Kalendereintrags ", example = "2022-10-13 14:00:00.000")
	public String kalenderEnde;

	/** ob der angemeldete Nutzer Schreibrecht auf dem Kalender hat */
	@Schema(required = true, description = "Schreibrecht des angemeldeten Nutzers", example = "true")
	public boolean darfSchreiben;

	/** ob der angemeldete Nutzer Leserecht auf dem Kalender hat */
	@Schema(required = true, description = "Leserecht des angemeldeten Nutzers", example = "true")
	public boolean darfLesen;

	/** ob der angemeldete Benutzer Besitzer des Kalenders ist, zu dem dieser Eintrag gehört */
	@Schema(required = true, description = "ob der angemeldete Benutzer Besitzer des Kalenders ist, zu dem dieser Eintrag gehört", example = "true")
	public boolean istBesitzer;

	/** der Typ des Kalendereintrags */
	@Schema(required = true, description = "der Typ des Kalendereintrags ", example = "VEVENT")
	public @NotNull String kalenderTyp = "VEVENT";
}

File: svws-webclient/src/core/ts/src/core/data/kalender/KalenderEintrag.ts
import { JavaObject } from '../../../java/lang/JavaObject';

export class KalenderEintrag extends JavaObject {

	/**
	 * ID des Kalendereintrags
	 */
	public id : string = "";

	/**
	 * ID des Kalenders
	 */
	public kalenderId : string = "";

	/**
	 *  URI der ICS-Repräsentation des Kalendereintrags
	 */
	public uid : string = "";

	/**
	 *  Versionskennzeichen des Kontaks
	 */
	public version : string = "";

	/**
	 * das serialisierte .ics dieses Kalendereintrags
	 */
	public data : string = "";

	/**
	 * der Startzeitpunkt dieses Kalendereintrags als SQL-Timestamp
	 */
	public kalenderStart : string | null = null;

	/**
	 * der Endzeitpunkt dieses Kalendereintrags als SQL-Timestamp
	 */
	public kalenderEnde : string | null = null;

	/**
	 * ob der angemeldete Nutzer Schreibrecht auf dem Kalender hat
	 */
	public darfSchreiben : boolean = false;

	/**
	 * ob der angemeldete Nutzer Leserecht auf dem Kalender hat
	 */
	public darfLesen : boolean = false;

	/**
	 * ob der angemeldete Benutzer Besitzer des Kalenders ist, zu dem dieser Eintrag gehört
	 */
	public istBesitzer : boolean = false;

	/**
	 * der Typ des Kalendereintrags
	 */
	public kalenderTyp : string = "VEVENT";


	public constructor() {
		super();
	}

	isTranspiledInstanceOf(name : string): boolean {
		return ['de.nrw.schule.svws.core.data.kalender.KalenderEintrag'].includes(name);
	}

	public static transpilerFromJSON(json : string): KalenderEintrag {
		const obj = JSON.parse(json);
		const result = new KalenderEintrag();
		if (typeof obj.id === "undefined")
			 throw new Error('invalid json format, missing attribute id');
		result.id = obj.id;
		if (typeof obj.kalenderId === "undefined")
			 throw new Error('invalid json format, missing attribute kalenderId');
		result.kalenderId = obj.kalenderId;
		if (typeof obj.uid === "undefined")
			 throw new Error('invalid json format, missing attribute uid');
		result.uid = obj.uid;
		if (typeof obj.version === "undefined")
			 throw new Error('invalid json format, missing attribute version');
		result.version = obj.version;
		if (typeof obj.data === "undefined")
			 throw new Error('invalid json format, missing attribute data');
		result.data = obj.data;
		result.kalenderStart = typeof obj.kalenderStart === "undefined" ? null : obj.kalenderStart === null ? null : obj.kalenderStart;
		result.kalenderEnde = typeof obj.kalenderEnde === "undefined" ? null : obj.kalenderEnde === null ? null : obj.kalenderEnde;
		if (typeof obj.darfSchreiben === "undefined")
			 throw new Error('invalid json format, missing attribute darfSchreiben');
		result.darfSchreiben = obj.darfSchreiben;
		if (typeof obj.darfLesen === "undefined")
			 throw new Error('invalid json format, missing attribute darfLesen');
		result.darfLesen = obj.darfLesen;
		if (typeof obj.istBesitzer === "undefined")
			 throw new Error('invalid json format, missing attribute istBesitzer');
		result.istBesitzer = obj.istBesitzer;
		if (typeof obj.kalenderTyp === "undefined")
			 throw new Error('invalid json format, missing attribute kalenderTyp');
		result.kalenderTyp = obj.kalenderTyp;
		return result;
	}

	public static transpilerToJSON(obj : KalenderEintrag) : string {
		let result = '{';
		result += '"id" : ' + '"' + obj.id! + '"' + ',';
		result += '"kalenderId" : ' + '"' + obj.kalenderId! + '"' + ',';
		result += '"uid" : ' + '"' + obj.uid! + '"' + ',';
		result += '"version" : ' + '"' + obj.version! + '"' + ',';
		result += '"data" : ' + '"' + obj.data! + '"' + ',';
		result += '"kalenderStart" : ' + ((!obj.kalenderStart) ? 'null' : '"' + obj.kalenderStart + '"') + ',';
		result += '"kalenderEnde" : ' + ((!obj.kalenderEnde) ? 'null' : '"' + obj.kalenderEnde + '"') + ',';
		result += '"darfSchreiben" : ' + obj.darfSchreiben + ',';
		result += '"darfLesen" : ' + obj.darfLesen + ',';
		result += '"istBesitzer" : ' + obj.istBesitzer + ',';
		result += '"kalenderTyp" : ' + '"' + obj.kalenderTyp! + '"' + ',';
		result = result.slice(0, -1);
		result += '}';
		return result;
	}

	public static transpilerToJSONPatch(obj : Partial<KalenderEintrag>) : string {
		let result = '{';
		if (typeof obj.id !== "undefined") {
			result += '"id" : ' + '"' + obj.id + '"' + ',';
		}
		if (typeof obj.kalenderId !== "undefined") {
			result += '"kalenderId" : ' + '"' + obj.kalenderId + '"' + ',';
		}
		if (typeof obj.uid !== "undefined") {
			result += '"uid" : ' + '"' + obj.uid + '"' + ',';
		}
		if (typeof obj.version !== "undefined") {
			result += '"version" : ' + '"' + obj.version + '"' + ',';
		}
		if (typeof obj.data !== "undefined") {
			result += '"data" : ' + '"' + obj.data + '"' + ',';
		}
		if (typeof obj.kalenderStart !== "undefined") {
			result += '"kalenderStart" : ' + ((!obj.kalenderStart) ? 'null' : '"' + obj.kalenderStart + '"') + ',';
		}
		if (typeof obj.kalenderEnde !== "undefined") {
			result += '"kalenderEnde" : ' + ((!obj.kalenderEnde) ? 'null' : '"' + obj.kalenderEnde + '"') + ',';
		}
		if (typeof obj.darfSchreiben !== "undefined") {
			result += '"darfSchreiben" : ' + obj.darfSchreiben + ',';
		}
		if (typeof obj.darfLesen !== "undefined") {
			result += '"darfLesen" : ' + obj.darfLesen + ',';
		}
		if (typeof obj.istBesitzer !== "undefined") {
			result += '"istBesitzer" : ' + obj.istBesitzer + ',';
		}
		if (typeof obj.kalenderTyp !== "undefined") {
			result += '"kalenderTyp" : ' + '"' + obj.kalenderTyp + '"' + ',';
		}
		result = result.slice(0, -1);
		result += '}';
		return result;
	}

}

export function cast_de_nrw_schule_svws_core_data_kalender_KalenderEintrag(obj : unknown) : KalenderEintrag {
	return obj as KalenderEintrag;
}
